javaslang/src/main/java/javaslang/collection/BitSet.java

```diff
@@ -0,0 +1,1024 @@
+                if (element < 0) {
+                    throw new IllegalArgumentException("bitset element must be >= 0");
+                }
+                setElement(copy, element);
+            });
+            return fromBitMaskNoCopy(copy);
+        }
+
+        @Override
+        public boolean contains(T t) {
+            final int element = toInt.apply(t);
+            if (element < 0) {
+                throw new IllegalArgumentException("bitset element must be >= 0");
+            }
+            final int index = element >> ADDRESS_BITS_PER_WORD;
+            return (getWord(index) & (1L << element)) != 0;
+        }
+
+        @Override
+        public BitSet<T> init() {
+            if (isEmpty()) {
+                throw new UnsupportedOperationException("init of empty TreeSet");
+            } else {
+                final long last = getWord(getWordsNum() - 1);
+                final int element = BITS_PER_WORD * (getWordsNum() - 1) + BITS_PER_WORD - Long.numberOfLeadingZeros(last) - 1;
+                return remove(fromInt.apply(element));
+            }
+        }
+
+        @Override
+        public Iterator<T> iterator() {
+            return Stream.range(0, getWordsNum() << ADDRESS_BITS_PER_WORD)
+                    .filter(i -> (getWord(i >> ADDRESS_BITS_PER_WORD) & (1L << i)) != 0)
+                    .map(fromInt)
+                    .iterator();
+        }
+
+        @Override
+        public int length() {
+            int len = 0;
+            for (int i = 0; i < getWordsNum(); i++) {
+                len += Long.bitCount(getWord(i));
+            }
+            return len;
+        }
+
+        @Override
+        public BitSet<T> take(long n) {
+            if (n <= 0) {
+                return createEmpty();
+            } else if (n >= length()) {
+                return this;
+            } else {
+                return createFromAll(iterator().take(n));
+            }
+        }
+
+        @Override
+        public BitSet<T> takeRight(long n) {
+            if (n <= 0) {
+                return createEmpty();
+            } else if (n >= length()) {
+                return this;
+            } else {
+                return createFromAll(iterator().takeRight(n));
+            }
+        }
+
+        @Override
+        public BitSet<T> remove(T t) {
+            if (contains(t)) {
+                final int element = toInt.apply(t);
+                final long[] copy = copyExpand(getWordsNum());
+                unsetElement(copy, element);
+                return fromBitMaskNoCopy(shrink(copy));
+            } else {
+                return this;
+            }
+        }
+
+        @Override
+        @SuppressWarnings("unchecked")
+        public BitSet<T> removeAll(Iterable<? extends T> elements) {
+            final Stream<Integer> source = Stream.ofAll(elements).map(toInt);
+            final long[] copy = copyExpand(getWordsNum());
+            source.forEach(element -> {
```

Comment:
minor: I think enhanced for loops are a bit cheaper (no lambda) and are just as readable